pkg/distributor/distributor.go

```diff
@@ -340,129 +376,188 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
+			if removeReplica {
+				removeLabel(d.limits.HAReplicaLabel(userID), &ts.Labels)
+			}
 
-		if len(ts.Labels) == 0 {
-			continue
-		}
+			for _, labelName := range d.limits.DropLabels(userID) {
+				removeLabel(labelName, &ts.Labels)
+			}
 
-		// We rely on sorted labels in different places:
-		// 1) When computing token for labels, and sharding by all labels. Here different order of labels returns
-		// different tokens, which is bad.
-		// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
-		// later in the validation phase, we ignore them here.
-		sortLabelsIfNeeded(ts.Labels)
+			if len(ts.Labels) == 0 {
+				continue
+			}
 
-		// Generate the sharding token based on the series labels without the HA replica
-		// label and dropped labels (if any)
-		key, err := d.tokenForLabels(userID, ts.Labels)
-		if err != nil {
-			return nil, err
-		}
+			// We rely on sorted labels in different places:
+			// 1) When computing token for labels, and sharding by all labels. Here different order of labels returns
+			// different tokens, which is bad.
+			// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
+			// later in the validation phase, we ignore them here.
+			sortLabelsIfNeeded(ts.Labels)
+
+			// Generate the sharding token based on the series labels without the HA replica
+			// label and dropped labels (if any)
+			key, err := d.tokenForLabels(userID, ts.Labels)
+			if err != nil {
+				return nil, err
+			}
 
-		validatedSeries, err := d.validateSeries(ts, userID)
+			validatedSeries, err := d.validateSeries(ts, userID)
 
-		// Errors in validation are considered non-fatal, as one series in a request may contain
-		// invalid data but all the remaining series could be perfectly valid.
-		if err != nil {
-			lastPartialErr = err
-		}
+			// Errors in validation are considered non-fatal, as one series in a request may contain
+			// invalid data but all the remaining series could be perfectly valid.
+			if err != nil {
+				lastPartialErr = err
+			}
 
-		// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
-		if validatedSeries == emptyPreallocSeries {
-			continue
+			// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
+			if validatedSeries == emptyPreallocSeries {
+				continue
+			}
+
+			seriesKeys = append(seriesKeys, key)
+			validatedTimeseries = append(validatedTimeseries, validatedSeries)
+			validatedSamples += len(ts.Samples)
 		}
+	}
 
-		keys = append(keys, key)
-		validatedTimeseries = append(validatedTimeseries, validatedSeries)
-		validatedSamples += len(ts.Samples)
+	if len(req.Metadata) > 0 {
+		for _, m := range req.Metadata {
+			key := d.tokenForMetadata(userID, m.MetricName)
+
+			err := d.validateMetadata(m.MetricMetadata, userID)
+
+			if err != nil {
+				// TODO: save the error to a partial error?
```

Comment:
We use `lastPartialErr` to tell the client whenever an error occurred or not, in my eyes this felt a bit misleading as we'd still push the good samples/metadata to the ingesters but tell the client that the request was not successful (with a 400 status).

I'm unsure on how to best proceed from here: should we override the `lastPartialErr` from the series and replace it with the metadata? Should we have two separate errors and handle them in the HTTP handler? In my eyes, the current approach is a bit misleading as we only return to the client one error even if multiple occurred. 

If it makes sense, I'd love to put together something that aggregates all the errors and sends them back to the client - it could be something like the last 10 errors that occurred while validating or something along those lines.

Reply:
> in my eyes this felt a bit misleading as we'd still push the good samples/metadata to the ingesters but tell the client that the request was not successful (with a 400 status).

Currently it's the only way we have to signal Prometheus that a partial error occurred. The error will be visible in logs but since it's a 4xx Prometheus will not retry to push it (which is what we want).

> I'd love to put together something that aggregates all the errors

I'm dubious about it. In case you're pushing 1K samples and they all generate an error (which may not be uncommon, depending on the error) you're going to join 1K errors in the response. I wouldn't do it.

> // TODO: save the error to a partial error?

I think we should save it, but I honestly believe that instead of `lastPartialErr` we should keep track of `firstPartialErr` and thus here we should save the error only if `firstPartialErr` is `nil`. Given should be an easy change, may you submit a PR to change `lastPartialErr` into `firstPartialErr` so we get feedback from other maintainers too, without further making the diff more complex here?

Reply:
> I'm dubious about it. In case you're pushing 1K samples and they all generate an error (which may not be uncommon, depending on the error) you're going to join 1K errors in the response. I wouldn't do it.

My writing suggested using the last 10 for that reason. I agree aggregating _every_ error is unfeasible. But the key takeaway to what I'm saying is that we signal, there could be > 1. which is not clear. 

>  think we should save it, but I honestly believe that instead of lastPartialErr we should keep track of firstPartialErr and thus here we should save the error only if firstPartialErr is nil. Given should be an easy change, may you submit a PR to change lastPartialErr into firstPartialErr so we get feedback from other maintainers too, without further making the diff more complex here?

Agree. Will do 👍

Reply:
Filed in https://github.com/cortexproject/cortex/pull/2383

Reply:
I'm going to resolve this comment given all conversation regarding this should happen in that PR.